<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
	<meta name="KeyWords"
			 content="PAL, phylogenetic analysis library, molecular phylogenetics,
					molecular evolution, phylogeny,
								tree reconstruction, maximum likelihood">
	<meta name="Content-Type" content="text/html; charset=iso-8859-1">
	<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
	<title>The PAL Project</title>
</head>
<body>

<table cellpadding="4" cellspacing="4" border="0" width="100%" height="100%">
		<tbody>
			<tr>
				<td valign="top">
					<table cellpadding="0" cellspacing="0" border="1" width="130" height = "500">
						<tbody>
							<tr>
								<td valign="top" bgcolor="#cccccc" align="center">
									<a href = "index.html">About</a>
									<br>
									<a href = "people.html"> People Involved </a>
									<br>
									<a href = "features.html">Features</a>
									<br>
									<a href = "user_interfaces.html">User Interfaces</a>
									<br>
									<b>Related Projects</b>
									<br>
									<br>
									<a href = "download.html">Downloads and Installation</a>
									<br>
									<a href = "history.html">History</a>
									<br>
									<br>
									<a href = "contributing.html">Contributing</a>
									<br>
									<a href = "acknowledgments.html">Acknowledgements</a>
									<br>
									<br>
									<a href = "online.html">Pal Online</a>
								</td>
							</tr>
						</tbody>
					</table>
				</td>
				<td valign="top">
	<!-- Start of normal block -->

<img src=pal.gif hspace=10 align=left ALT="PAL icon">
<strong><font size="+2">PAL: Phylogenetic Analysis Library</font></strong>

<br>
<br>
<strong><em>
<center>Related (Java) Projects</center>
</em></strong>
<p>
To our knowledge, PAL is the only  project that aims at
providing a collaborative Java library for molecular evolution and phylogenetics.
<p>
Similar "private" projects are, however,
currently undertaken, e.g.,
by Andrew Rambaut and Mike Charleston
<a href="http://evolve.zoo.ox.ac.uk/projects/BBSRC-Nautilus/Nautilus.html">(their Nautilus project)</a>
or by the <a href="http://evolution.genetics.washington.edu/">Felsenstein lab</a> (both in C++).

<p>
Other Java projects in molecular evolution and bioinformatics include:
<ul>

<li>
<a href="http://mesquiteproject.org">Mesquite</a>,
a modular system for evolutionary analysis is developed by Wayne Maddison
and David Maddison.  Mesquite is designed to allow  third-party components
and plug-in modules and provides a nice graphical user interface.

</li>

<li>
The <a href="http://www.biojava.org">BioJava project</a> is collaborative
project that aims at creating a Java library for general bioinformatics
applications (BLAST, CORBA etc.).
</li>

<li>
Christian Zmasek's program <a href="http://www.genetics.wustl.edu/eddy/forester/">FORESTER</a>
(BSD license) is a Java project aimed at phylogenomics.
<a href="http://www.genetics.wustl.edu/eddy/atv/">ATV</a> is part of
FORESTER and is a useful tree viewer that can also displays trees written
in the NHX format (an extended NH format).
</li>

<li>
John Brzustowski distributes
<a href ="http://www.biology.ualberta.ca/jbrzusto/cluster.php">qclust</a>,
a set of Java classes implementing a variety of clustering methods, including
UPGMA and neighbor-joining.
</li>

<li>
Ed Buckler has developed a number of Java applications
for phylogenetic and linkage analysis
(e.g.,  <a
href="http://brooks.statgen.ncsu.edu/buckler/bioinformatics.html">Phylogeographer and TASSEL</a>
).  Ed is now a co-developer of PAL.
</li>

<li>
Don Gilbert distributes a number of his Java programs in his IUBio Archive,
for example the sequence editor
<a href ="http://iubio.bio.indiana.edu/soft/molbio/java/apps/seqpup/">SeqPup</a>
and the tree drawing program
<a href ="http://iubio.bio.indiana.edu/soft/molbio/java/apps/trees/">Phylodendron</a>.
</li>
<li>David Posada and collaborators have written
<a href="http://bioag.byu.edu/zoology/crandall_lab/geodis.htm">GEODIS</a>,
a Java program for cladistic nested analysis of the geographical distribution
of genetic haplotypes,
and
<a href="http://bioag.byu.edu/zoology/crandall_lab/tcs.htm">TCS</a>,
a Java program to estimate gene genealogies
using statistical parsimony.
</li>

</ul>
<p>
Finally, the general issue of scientific programming in Java is discussed, e.g., on the
<a href="http://math.nist.gov/javanumerics/">Java Numerics</a>
web page.

	<!-- End of normal block -->


				</td>
			</tr>

	</tbody>
</table>
	<br>

</body>
</html>
